backupccl: reset job level retry counter if backup progress has advanced

lyang24 opened the pull request:
This commit hardens the backup's retry loop by resetting the retry counter on
meaningful progress. In details, The retry loop will be reseted if the backup
job encountered a retriable error and it have made at least 1% more progress
than the previous run.

Fixes: https://github.com/cockroachdb/cockroach/issues/119691

Release note: None

lyang24 committed:
backupccl: reset job level retry counter if backup progress has advanced

This commit hardens the backup's retry loop by resetting the retry counter on
meaningful progress. In details, The retry loop will be reseted if the backup
job encountered a retriable error and it have made at least 1% more progress
than the previous run.

Fixes: https://github.com/cockroachdb/cockroach/issues/119691

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119804)
<!-- Reviewable:end -->

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1638,63 @@ func TestRestoreRetryProcErr(t *testing.T) {
 
+// TestBackupJobRetryReset tests that the job level retry counter
+// resets after the backup made progresses. To do so, the test does the following:
+// 1. Intercept the backup job after a backup chuck completes and returns a retryable
```
s/chuck/chunk/r

msbutler commented on pkg/sql/exec_util.go:
```diff
@@ -1626,6 +1626,10 @@ type ExecutorTestingKnobs struct {
 	// transaction deadlines even in the system tenant.
 	ForceSQLLivenessSession bool
+
+	// BeforeBackupResetRetry is called before backup loop encountering a retryable
```
s/encountering/encounters/r

msbutler replied:
you can remove "it returns ..."

msbutler commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -860,6 +869,29 @@ func (b *backupResumer) Resume(ctx context.Context, execCtx interface{}) error {
 		}
+		// Re-load the job in order to update our progress object, which
+		// may have been updated since the flow started.
+		reloadedJob, reloadErr := p.ExecCfg().JobRegistry.LoadClaimedJob(ctx, b.job.ID())
```
In general, we shouldn't need to reload the job from the job coordinator (i.e the node running resume). Check out how we avoid a job re-load in restore [here.](https://github.com/msbutler/cockroach/blob/dcd1638571e12f80da01130374bc2ffe58a14697/pkg/ccl/backupccl/restore_job.go#L212)

msbutler replied:
the import code is provides a nice template as well https://github.com/cockroachdb/cockroach/pull/119708/commits/5e6ce265363538a21d49afb91393f52dc76a75e3

msbutler reviewed (commented):
Off to a good start! I have yet to look at the test. Will do that after my next set of meetings.

lyang24 commented:
*[`pkg/ccl/backupccl/backup_job.go` line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1511395551:-Ns9tdroBgygjmneolWu:b-2fgcea) ([raw file](https://github.com/cockroachdb/cockroach/blob/adcbf7967e4dd2f01f406f612e882273942831eb/pkg/ccl/backupccl/backup_job.go#L874)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

the import code is provides a nice template as well https://github.com/cockroachdb/cockroach/pull/119708/commits/5e6ce265363538a21d49afb91393f52dc76a75e3
</blockquote></details>

Appreciates the feedbacks and this part is not clear to me. I think the import code uses job reload and the backup progress is not saved on backup progress's proto message.

In details:

I copied the reload from the import code [here]  (https://github.com/lyang24/cockroach/blob/0ac60589fefd29322cc5fe4e4eb3e7c8135222fc/pkg/sql/importer/import_job.go#L1302)

from the[restore example](https://github.com/msbutler/cockroach/blob/dcd1638571e12f80da01130374bc2ffe58a14697/pkg/ccl/backupccl/restore_job.go#L212) the checkpoint is saved on restore progress. However, the backup progress [defined](https://github.com/lyang24/cockroach/blob/9d71a8e805d505ed6fd4ed31eb3c3aa3bf3b1caf/pkg/jobs/jobspb/jobs.proto#L356-L358) in proto looks empty.


<!-- Sent from Reviewable.io -->

lyang24 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119804)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler)

___
*[`pkg/sql/exec_util.go` line 1630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1511386460:-NsBv_v8Cp754tYuu0-F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/adcbf7967e4dd2f01f406f612e882273942831eb/pkg/sql/exec_util.go#L1630)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

you can remove "it returns ..."
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1643 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1511383084:-NsBvauZBk3KhLiCF73b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/adcbf7967e4dd2f01f406f612e882273942831eb/pkg/ccl/backupccl/backup_test.go#L1643)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

s/chuck/chunk/r
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1638,63 @@ func TestRestoreRetryProcErr(t *testing.T) {
 }
 
+// TestBackupJobRetryReset tests that the job level retry counter
+// resets after the backup made progresses. To do so, the test does the following:
```
s/made//r

msbutler replied:
s/made//r

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1638,63 @@ func TestRestoreRetryProcErr(t *testing.T) {
 
+// TestBackupJobRetryReset tests that the job level retry counter
+// resets after the backup made progresses. To do so, the test does the following:
+// 1. Intercept the backup job after a backup chunk completes and returns a retryable
```
s/returns/return/r

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1638,63 @@ func TestRestoreRetryProcErr(t *testing.T) {
+// TestBackupJobRetryReset tests that the job level retry counter
+// resets after the backup made progresses. To do so, the test does the following:
+// 1. Intercept the backup job after a backup chunk completes and returns a retryable
+// error until the retry reached desired count.
```
s/retry reached /retry counter reaches the/r

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1638,63 @@ func TestRestoreRetryProcErr(t *testing.T) {
+// resets after the backup made progresses. To do so, the test does the following:
+// 1. Intercept the backup job after a backup chunk completes and returns a retryable
+// error until the retry reached desired count.
+// 2. Added a new testing know BeforeBackupResetRetry to simulate backup job has made
```
there are few typos in this sentence. please do reread comments before submitting a PR for review.

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1638,63 @@ func TestRestoreRetryProcErr(t *testing.T) {
+		JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+	}
+	params.ServerArgs = base.TestServerArgs{Knobs: knobs}
+	// Best effort to set up backup jobs with spans > 1.
```
nit: to ensure we see several progress messages, you can reduce the range size for this table via zone config cmds. https://www.cockroachlabs.com/docs/stable/alter-range

How long does this test take to run?

msbutler replied:
another way to ensure you produce multiple backup spans is to backup multiple tables in the database.

msbutler commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -860,6 +869,29 @@ func (b *backupResumer) Resume(ctx context.Context, execCtx interface{}) error {
+			curProgress := reloadedJob.FractionCompleted()
+			var retryForTest bool
+			if p.ExecCfg().TestingKnobs.BeforeBackupResetRetry != nil {
+				retryForTest = p.ExecCfg().TestingKnobs.BeforeBackupResetRetry()
```
so this retryForTest testing knob is essentially removing  testing coverage around the logic you've added for checking current and previous progress. Could you try removing this testing knob (in a new commit, to prevent erasing the good work you've already done), and adding a few more coordination mechanisms to the unit test? Check out how `TestRestoreJobRetryReset` has done this.

msbutler commented:
@lyang24 you were right about the import code! you can keep the job reload logic in here.

lyang24 committed:
backupccl: improve testing on backup job retry reset

This commit improves TestBackupJobRetryReset's reliability and coverage area.

Epic: None

Release note: None

lyang24 reviewed (commented):
Thanks for reviewing

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119804#-:-NsS4MbR507sa5W0E6Dc:b-lr058f)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @msbutler)

___
*[`pkg/ccl/backupccl/backup_job.go` line 885 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514608065:-NsS2rqg-Mf_XdpXWuxv:blslexg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_job.go#L885)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

so this retryForTest testing knob is essentially removing  testing coverage around the logic you've added for checking current and previous progress. Could you try removing this testing knob (in a new commit, to prevent erasing the good work you've already done), and adding a few more coordination mechanisms to the unit test? Check out how `TestRestoreJobRetryReset` has done this. 
</blockquote></details>

This is great, the previous test is unreliable sometimes it hangs at `jobutils.WaitForJobToSucceed` call. So I went all in with mirroring what `TestRestoreJobRetryReset` has done. I tested a couple of times, it took around 1s to run and results are reliable, it hit the retry reset exactly and eventually fails with retry exhausted.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1642 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514558148:-NsS3wyg1ZR8LZ1-Xnac:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_test.go#L1642)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

s/made//r
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1642 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514559336:-NsS3xfEC5TFs_8uugcA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_test.go#L1642)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

s/made//r
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514559972:-NsS4-_N35wYYphL5OZL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_test.go#L1643)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

s/returns/return/r
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1644 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514561327:-NsS40JyEYUf_xCUi2G0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_test.go#L1644)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

s/retry reached /retry counter reaches the/r
</blockquote></details>

Done.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1645 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514563705:-NsS40cj1x5ZOd9rXTXk:bhr58zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_test.go#L1645)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

there are few typos in this sentence. please do reread comments before submitting a PR for review.
</blockquote></details>

my bad will do.
___
*[`pkg/ccl/backupccl/backup_test.go` line 1687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1514567372:-NsS46b328GP5kfZuX6Y:bsb8vqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62f0f24e57b4434065b121117f27ea101ca8a13/pkg/ccl/backupccl/backup_test.go#L1687)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

another way to ensure you produce multiple backup spans is to backup multiple tables in the database.
</blockquote></details>

I went with create table method on this one it works well. the new test took 1s to run.


<!-- Sent from Reviewable.io -->

msbutler commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -1637,6 +1637,75 @@ func TestRestoreRetryProcErr(t *testing.T) {
+	sqlDB.QueryRow(t, `BACKUP DATABASE data INTO $1 WITH DETACHED`, localFoo).Scan(&backupJobId)
+	testutils.SucceedsSoon(t, func() error {
+		jobProgress := jobutils.GetJobProgress(t, sqlDB, backupJobId)
+		if jobProgress.GetFractionCompleted() < 1 {
```
There's also a bug on master with `FractionCompleted`. You rebase your PR on this https://github.com/cockroachdb/cockroach/pull/120204

lyang24 reviewed (commented):
Thanks for catching the condition bug.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119804#-:-Nsqlp-s-B3EduRQv7O3:bedu89q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @msbutler)

___
*[`pkg/ccl/backupccl/backup_test.go` line 1698 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119804#gh-1519735238:-NsqlYDuBfkbepAusiOw:b-z8dbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/53037bab59bda90aa5fb2c8417dc7d162abdf35b/pkg/ccl/backupccl/backup_test.go#L1698)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

There's also a bug on master with `FractionCompleted`. You rebase your PR on this https://github.com/cockroachdb/cockroach/pull/120204
</blockquote></details>

fixed the condition and rebased on master that included #120204 


<!-- Sent from Reviewable.io -->

msbutler reviewed (approved):
Looks great! Thanks for doing this and being patient with all my nits!
